crates/glaredb_core/src/functions/scalar/builtin/binary/shl.rs

```diff
@@ -0,0 +1,145 @@
+        ),
+        RawScalarFunction::new(
+            &Signature::new(&[DataTypeId::Int32, DataTypeId::Int32], DataTypeId::Int32),
+            &Shl::<PhysicalI32>::new(&DataType::Int32),
+        ),
+        RawScalarFunction::new(
+            &Signature::new(&[DataTypeId::Int64, DataTypeId::Int64], DataTypeId::Int64),
+            &Shl::<PhysicalI64>::new(&DataType::Int64),
+        ),
+        RawScalarFunction::new(
+            &Signature::new(
+                &[DataTypeId::Int128, DataTypeId::Int128],
+                DataTypeId::Int128,
+            ),
+            &Shl::<PhysicalI128>::new(&DataType::Int128),
+        ),
+        RawScalarFunction::new(
+            &Signature::new(&[DataTypeId::UInt8, DataTypeId::UInt8], DataTypeId::UInt8),
+            &Shl::<PhysicalU8>::new(&DataType::UInt8),
+        ),
+        RawScalarFunction::new(
+            &Signature::new(
+                &[DataTypeId::UInt16, DataTypeId::UInt16],
+                DataTypeId::UInt16,
+            ),
+            &Shl::<PhysicalU16>::new(&DataType::UInt16),
+        ),
+        RawScalarFunction::new(
+            &Signature::new(
+                &[DataTypeId::UInt32, DataTypeId::UInt32],
+                DataTypeId::UInt32,
+            ),
+            &Shl::<PhysicalU32>::new(&DataType::UInt32),
+        ),
+        RawScalarFunction::new(
+            &Signature::new(
+                &[DataTypeId::UInt64, DataTypeId::UInt64],
+                DataTypeId::UInt64,
+            ),
+            &Shl::<PhysicalU64>::new(&DataType::UInt64),
+        ),
+        RawScalarFunction::new(
+            &Signature::new(
+                &[DataTypeId::UInt128, DataTypeId::UInt128],
+                DataTypeId::UInt128,
+            ),
+            &Shl::<PhysicalU128>::new(&DataType::UInt128),
+        ),
+    ],
+};
+
+#[derive(Debug, Clone, Copy)]
+pub struct Shl<S> {
+    return_type: &'static DataType,
+    _s: PhantomData<S>,
+}
+
+impl<S> Shl<S> {
+    pub const fn new(return_type: &'static DataType) -> Self {
+        Shl {
+            return_type,
+            _s: PhantomData,
+        }
+    }
+}
+
+impl<S> ScalarFunction for Shl<S>
+where
+    S: MutableScalarStorage,
+    S::StorageType: std::ops::Shl<Output = S::StorageType> + Sized + Copy,
+{
+    type State = ();
+
+    fn bind(&self, inputs: Vec<Expression>) -> Result<BindState<Self::State>> {
+        Ok(BindState {
+            state: (),
+            return_type: self.return_type.clone(),
+            inputs,
+        })
+    }
+
+    fn execute(_state: &Self::State, input: &Batch, output: &mut Array) -> Result<()> {
+        let sel = input.selection();
+        let a = &input.arrays()[0];
+        let b = &input.arrays()[1];
+
+        BinaryExecutor::execute::<S, S, S, _>(
+            a,
+            sel,
+            b,
+            sel,
+            OutBuffer::from_array(output)?,
+            |&a, &b, buf| buf.put(&(a << b)),
+        )
```

Comment:
```suggestion
        BinaryExecutor::execute::<S, PhysicalI32, S, _>(
            a,
            sel,
            b,
            sel,
            OutBuffer::from_array(output)?,
            |&a, &b, buf| buf.put(&(a << b)),
        )
```

After making the adjustment to the signatures to use DataType::Int32.